apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/CloudTraceExtension.cs

```diff
@@ -112,32 +110,28 @@ public static IApplicationBuilder UseGoogleTrace(this IApplicationBuilder app)
             var serviceOptions = new TraceServiceOptions();
             setupAction(serviceOptions);
 
-            var client = serviceOptions.Client ?? TraceServiceClient.Create();
-            var options = serviceOptions.Options ?? TraceOptions.Create();
-            var traceFallbackPredicate = serviceOptions.TraceFallbackPredicate ?? TraceDecisionPredicate.Default;
-            var projectId = Project.GetAndCheckProjectId(serviceOptions.ProjectId);
+            services.AddGoogleTrace(SetupCommonOptions);
+            void SetupCommonOptions(Common.TraceServiceOptions commonOptions)
+            {
+                commonOptions.ProjectId = serviceOptions.ProjectId;
+                commonOptions.Options = serviceOptions.Options;
+                commonOptions.Client = serviceOptions.Client;
+            }
 
-            var consumer = ManagedTracer.CreateConsumer(client, options);
-            var tracerFactory = ManagedTracer.CreateFactory(projectId, consumer, options);
+            var traceFallbackPredicate = serviceOptions.TraceFallbackPredicate ?? TraceDecisionPredicate.Default;
 
             // We use TryAdd... here to allow user code to inject their own trace context provider
             // and matching trace context response propagator. We use Google trace header otherwise.
             services.TryAddScoped<ITraceContext>(GoogleTraceHeaderContextProvider);
             services.TryAddSingleton<Action<HttpResponse, ITraceContext>>(GoogleTraceHeaderResponsePropagator);
 
-            services.AddSingleton(tracerFactory);
             // Obsolete: Adding this for backwards compatibility in case someone is using the old factory type.
             // The new and prefered factory type is Func<ITraceContext, IManagedTracer>.
-            services.AddSingleton<Func<TraceHeaderContext, IManagedTracer>>(tracerFactory);
+            services.AddSingleton<Func<TraceHeaderContext, IManagedTracer>>(sp => sp.GetRequiredService<Func<ITraceContext, IManagedTracer>>());
+
             services.AddHttpContextAccessor();
-            services.AddSingleton(ManagedTracer.CreateDelegatingTracer(ContextTracerManager.GetCurrentTracer));
             services.AddTransient<ICloudTraceNameProvider, DefaultCloudTraceNameProvider>();
 
-            // On .Net Standard 2.0 or higher, we can use the System.Net.Http.IHttpClientFactory defined in Microsoft.Extensions.Http,
-            // for which we need a DelagatingHandler with no InnerHandler set. This is the recommended way.
-            // It should be registered as follows.
-            services.AddTransient(UnchainedTraceHeaderPropagatingHandlerFactory);
```

Comment:
There's a lot of services being removed here... are we confident this won't break folks?